### DISCUSSION TITLE
分析下异常流量来源

### DISCUSSION THREAD

**Discussion by pencilnav** [XIU2/CloudflareSpeedTest#452] [讨论]
在更换域名到pencilfiles.xxx.xxx之前，我一直使用的是testfiles.xxx.xxx的链接对外服务，并且基本没有做任何限制 （除Cloudflare自己的规则之外）。
但我发现，当下载链接没有做任何限制的时候，其实有很大一部分流量都不是来自本项目的（基本上50%-60%），严重能在只有300个unique visitors的情况下一天300k请求数暴刷3T流量。这些IP其实都有共同点，即：大多为Cloudflare CDN IP，以及奇奇怪怪的User Agent。

当时我发现此问题后，就在Redirect Rule里设置了一条规则，将所有非中国IP全部302 redirect至999857.xyz（我的主站）的一个页面，然后利用Google Analytics和Cloudflare Analytics分析流量（暂不放图）。同时也借助Censys查看这些流量的源URL究竟是啥。

然后，我就看到了这个：
([censys链接](https://stlink.cf/xcQ7Zz))
![](https://static.999857.xyz/penciln/production/media/image-eaa5ff039901a66a289ce71edb8a19b7.png) (1569个指向我主站的主机？？？）
![](https://static.999857.xyz/penciln/production/media/image-6b6058a1ff7a756214a81461345821c1.png)

足足有1500多个指向我主站的链接，就算去掉我绑定过的所有DNS记录也有1400多个。而且其中也有大量的workers.dev域名，经测试全部都是反代我的文件的。

再来看User Agent：
![](https://static.999857.xyz/penciln/production/media/image-cdb4d3b5a06d71c97503d581c65d74c9.png)
![](https://static.999857.xyz/penciln/production/media/image-8cf52ac58da02134488de2cf0ab23613.png)
也是平时不怎么常见的。

我也在WAF里看到了好几个同IP在一个小时内被拦截上千次的记录。

目前的话，感觉也没什么好的解决办法。如果有遇到类似情况的链接提供者也可以在下面反馈下你的情况，看看如何解决。
临时解决也有。 1、在WAF内设置白名单，只允许本项目的[UserAgent](https://github.com/XIU2/CloudflareSpeedTest/blob/49992a763eb48310c8f649fec10be0a7e7c4b613/task/download.go#L129)下载。2、仅限中国IP访问（目前我的链接全部都是这样）。当然，我也有把以上二者同时部署的考虑。

**Comment by pencilnav** [comment]
@XIU2 如果可以的话，推荐在你的重定向链接cf.xiu2.xyz添加一条WAF rule，可以防止一些滥用。
![](https://static.999857.xyz/penciln/production/media/image-ca8c422178c86d2ab8b38a992af51f66.png)

表达式如下：
```
(http.host eq "cf.xiu2.xyz" and http.user_agent ne "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.80 Safari/537.36")
```

**Reply by XIU2** [reply]
我其实很早就做了限制了，只允许了部分地区的 IP 访问（比如国内、港澳台这些）。
